cache/network/api.go

```diff
@@ -0,0 +1,230 @@
+package network
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+
+	"github.com/hashicorp/go-retryablehttp"
+)
+
+const maxKeyLength = 512
+const maxKeyCount = 8
+
+type prepareUploadRequest struct {
+	CacheKey           string `json:"cache_key"`
+	ArchiveFileName    string `json:"archive_filename"`
+	ArchiveContentType string `json:"archive_content_type"`
+	ArchiveSizeInBytes int64  `json:"archive_size_in_bytes"`
+}
+
+type prepareUploadResponse struct {
+	ID            string            `json:"id"`
+	UploadMethod  string            `json:"method"`
+	UploadURL     string            `json:"url"`
+	UploadHeaders map[string]string `json:"headers"`
+}
+
+type restoreResponse struct {
+	URL string `json:"url"`
+}
+
+type apiClient struct {
+	httpClient  *retryablehttp.Client
+	baseURL     string
+	accessToken string
+}
+
+func newAPIClient(client *retryablehttp.Client, baseURL string, accessToken string) apiClient {
+	return apiClient{
+		httpClient:  client,
+		baseURL:     baseURL,
+		accessToken: accessToken,
+	}
+}
+
+func (c apiClient) prepareUpload(requestBody prepareUploadRequest) (prepareUploadResponse, error) {
+	url := fmt.Sprintf("%s/upload", c.baseURL)
+
+	body, err := json.Marshal(requestBody)
+	if err != nil {
+		return prepareUploadResponse{}, err
+	}
+
+	req, err := retryablehttp.NewRequest("POST", url, body)
```

Comment:
NOTE: `http.MethodPost`

Reply:
Yes, I considered using that, but the `retryablehttp` package also adds other useful bits, such as automatically computing the `Content-Length` header. And there is only a dedicated `retryablehttp.Get()` method, no `retryablehttp.Post()`.

Reply:
Ahh, sorry for the confusion, I meant, instead of hardcoding the `"POST"`, stdlib has a const for that: `http.MethodPost`

Reply:
Whoops. I fixed it now, take a look!